internal/provider/data_source_aws_network_peering.go

```diff
@@ -107,31 +108,63 @@ func dataSourceAwsNetworkPeeringRead(ctx context.Context, d *schema.ResourceData
 	peeringID := d.Get("peering_id").(string)
 	waitForActive := d.Get("wait_for_active_state").(bool)
 
+	// Query for the peering.
 	log.Printf("[INFO] Reading network peering (%s)", peeringID)
 	peering, err := clients.GetPeeringByID(ctx, client, peeringID, hvnID, loc)
 	if err != nil {
 		return diag.Errorf("unable to retrieve network peering (%s): %v", peeringID, err)
 	}
 
-	if waitForActive && peering.State != networkmodels.HashicorpCloudNetwork20200907PeeringStateACTIVE {
-		peering, err = clients.WaitForPeeringToBeActive(ctx, client, peering.ID, hvnID, loc, peeringCreateTimeout)
-		if err != nil {
-			return diag.FromErr(err)
-		}
+	// Set the globally unique id of this peering in state.
+	link := newLink(peering.Hvn.Location, PeeringResourceType, peering.ID)
+	url, err := linkURL(link)
+	if err != nil {
+		return diag.FromErr(err)
 	}
+	d.SetId(url)
 
-	// Network peering found, update resource data.
+	// Store resource data.
 	if err := setAwsPeeringResourceData(d, peering); err != nil {
 		return diag.FromErr(err)
 	}
 
-	// Set the globally unique id of this peering in the state.
-	link := newLink(peering.Hvn.Location, PeeringResourceType, peering.ID)
-	url, err := linkURL(link)
-	if err != nil {
-		return diag.FromErr(err)
+	// Skip waiting.
+	if !waitForActive || peering.State == models.HashicorpCloudNetwork20200907PeeringStateACTIVE {
+		return nil
 	}
-	d.SetId(url)
 
-	return nil
+	// If it's in a state where it could later become ACTIVE, wait.
+	waitState := false
+	for _, state := range clients.WaitForPeeringToBeActiveStates {
+		if state == string(peering.State) {
+			waitState = true
+			break
+		}
+	}
+
+	// If it's not in a state that we should wait on, issue a warning and bail.
+	if !waitState {
+		return []diag.Diagnostic{{
+			Severity: diag.Warning,
+			Summary:  fmt.Sprintf("Peering is in an unexpected state, connections may fail: %q", string(peering.State)),
+			Detail:   "Expected a CREATING, PENDING_ACCEPTANCE, ACCEPTED, or ACTIVE state",
+		}}
+	}
+
+	// Store resource data again, updating Peering state.
+	var result []diag.Diagnostic
+	peering, err = clients.WaitForPeeringToBeActive(ctx, client, peering.ID, hvnID, loc, peeringCreateTimeout)
```

Comment:
What happens if we transition from PENDING -> FAILED? The previous state will say that it's PENDING and then a destroy will try to read it above (without refresh false) and error out ?

Reply:
1. tf apply with PENDING -> FAILED
    - we still error out here in the read
2. tf apply again:
    - if we reached this again, we'll throw a warning because we won't wait. In practice I don't expect this to happen, because we'll (normally) have a resource that will see the `FAILED` state, mark the peering as tainted, and create a new peering in its place
3. tf destroy:
    - if we read this we'll just get the warning